README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# passt

<!-- badges: start -->

[![Build
Status](https://travis-ci.org/johannes-titz/passt.svg?branch=master)](https://travis-ci.org/johannes-titz/passt)
[![Codecov test
coverage](https://codecov.io/gh/johannes-titz/passt/branch/master/graph/badge.svg)](https://codecov.io/gh/johannes-titz/passt?branch=master)
[![CRAN
status](https://www.r-pkg.org/badges/version/passt)](https://CRAN.R-project.org/package=passt)
[![status](https://joss.theoj.org/papers/17dd34a44194e85822195179079e606c/status.svg)](https://joss.theoj.org/papers/17dd34a44194e85822195179079e606c)
<!-- badges: end -->

The *passt* package is an R implementation of the Probability ASSociator
Time (PASS-T) model, an artificial neural network designed to explain
how humans make judgments of frequency and duration (Titz & Sedlmeier,
2019). The package was developed with two purposes in mind: (1) to
provide a simple way to reproduce simulation results in judgments of
frequency and duration as described in Titz & Sedlmeier (2019), and (2)
to explore the PASS-T model by allowing users to run their own
individual simulations. The package is targeted at cognitive
psychologists who study judgments of frequency and duration, memory,
heuristics and artificial neural networks.

The general idea of the original PASS model (Sedlmeier, 1999, 2002) is
that information about the frequency of events is naturally incorporated
in artificial neural networks. If an object is presented repeatedly, the
weights of the network change systematically. This way, the network is
able to deduce the frequency of occurrence of events based on the final
weights. Put simply, the artificial neural network produces a higher
activation the more often a stimulus is presented.

As an extension of the PASS model, PASS-T is also able to process the
presentation time of the stimuli. One can define how often and how long
different objects are presented and test whether the network is able to
make valid judgments of frequency and duration in retrospect. The
principal aim of PASS-T is to explain empirical results that are found
in studies with humans (Titz et al., 2019). Specifically, two empirical
results are quite robust: (1) the correlation between exposure
*frequency* and judgment is usually larger than between exposure
*duration* and judgment (sometimes termed *frequency primacy*); and (2)
the amount of attention that participants pay to the stimuli moderates
effect sizes. These findings have been of some interest in cognitive
psychology in the last decade (Betsch, Glauer, Renkewitz, Winkler, &
Sedlmeier, 2010; Titz et al., 2019; Titz & Sedlmeier, 2019; Winkler,
2009; Winkler, Glauer, Betsch, & Sedlmeier, 2015), although its roots go
back half a century (Cooper & Pantle, 1967; Hintzman, 1970; Hintzman,
Summers, & Block, 1975) and touch upon different research areas such as
memory, heuristics, and perception of magnitudes. PASS-T can be seen as
a summary of some of this research, cast into a formal model. The R
package *passt* is the first concrete implementation of PASS-T to make
the model available to a wider audience.

## Installation

You can install the released version of passt from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("passt")
```

You can install the development version from
[GitHub](https://github.com/) with devtools (and vignettes build, this
takes a couple of seconds):

``` r
devtools::install_github("johannes-titz/passt", build_vignettes = TRUE)
```

## Using *passt*

There are only two functions of *passt* that you will need to use. The
first function is *run\_sim*, which runs several simulations with
specific parameters and returns the final output activation for each
input pattern. The second is *run\_exp*, which aggregates the data and
gives effect sizes for each simulation run. Let us first look at
*run\_sim*.

Since PASS-T extends the PASS model, PASS-T should be able to produce
all the results that PASS can produce. The simplest PASS simulation is a
type of counter: an artificial neural network sensitive only to
frequency information.

### A simple counter

``` r
library(passt)
set.seed(20191015)
```

``` r
sim1 <- run_sim(patterns = diag(10), frequency = 1:10, duration = 10:1,
                lrate_onset = 0.05, lrate_drop_time = 2, lrate_drop_perc = 0)
```

Some explanation is necessary: *patterns* is a matrix of input patterns
that the network will process. Here we will use orthogonal stimuli to
avoid any interference on the input level. The function *diag* will
create an identity matrix so that every input pattern has only one
active unit. This active unit will be unique for each pattern. You can
also use other stimulus patterns as long as the activation for each unit
is either 0 or 1.

Next, we specify how often the ten patterns should be shown and the
duration of each presentation. The first pattern will be shown once for
10 s, the second pattern will be shown twice for 9 s each (a total
duration of 18 s), and so on. You can also use other arrangements for a
simple counter.

Finally, we set the learning rate parameters: the initial learning rate
(*lrate\_onset*) is 0.05, but after 2 s (*lrate\_drop\_time*) it will
drop to 0 (*lrate\_drop\_perc*). This parameter selection is crucial for
a counter because only the first two seconds of a stimulus presentation
are processed by the network.

The function *run\_sim* returns a list with three elements, of which
*output* is the most interesting. Note that, by default, 100 simulations
are run (you can change this with the parameter *n\_runs*). We will only
look at the first couple of simulations:

``` r
head(sim1$output)
#>           [,1]      [,2]      [,3]      [,4]      [,5]     [,6]     [,7]     [,8]     [,9]    [,10]
#> [1,] 0.8134030 0.8623727 0.9090286 0.9526607 0.9893114 1.027876 1.065307 1.093939 1.125975 1.160127
#> [2,] 0.8161057 0.8618859 0.9028135 0.9479106 0.9905695 1.024340 1.062587 1.098260 1.132112 1.163416
#> [3,] 0.8117953 0.8602871 0.9049667 0.9490695 0.9895403 1.027940 1.061659 1.097431 1.134466 1.162845
#> [4,] 0.8148240 0.8633805 0.9009140 0.9490142 0.9884823 1.026991 1.061192 1.100671 1.128499 1.166032
#> [5,] 0.8114915 0.8611520 0.9011725 0.9449056 0.9886680 1.027973 1.067783 1.103455 1.132242 1.161158
#> [6,] 0.8177491 0.8598771 0.9061970 0.9493045 0.9831946 1.031035 1.059464 1.102026 1.132638 1.158515
```

The *output* gives the sum of the activity of all output units for the
specific patterns for each simulation run. Each row corresponds with one
simulation and each column with one input pattern. Sometimes the output
activation is referred to as the *memory strength* in an abstract sense.
One can easily see that increasing the presentation frequency
(i.e. going from the first to the tenth column) results in a higher
*memory strength* (activation).

### Simulating the moderating role of attention

The PASS-T model was used to explain how attention moderates effect
sizes in judgments of frequency and duration (Titz & Sedlmeier, 2019). A
short example for a setup could look like this:

``` r
duration <- c(4, 2, 1, 8, 4, 2, 12, 6, 3)
frequency <- c(2, 4, 8, 2, 4, 8, 2, 4, 8)
lrate_drop_perc <- seq(0, 1, 0.04)
sim4 <- lapply(lrate_drop_perc, function(x) 
  run_exp(frequency, duration, 0.05, 2, x, diag(9), 30, 0.1))
```

The *lapply* function goes through each *lrate\_drop\_perc* value and
runs a simulation. Note that we reduced the number of simulation runs to
30 to keep the computation time within reasonable limits. Since we now
have a list of simulations, we need to convert it into a data frame and
then add the information about the drop of the learning parameter:

``` r
sim4 <- plyr::ldply(sim4, "data.frame")
sim4 <- cbind(sim4, lrate_drop_perc)
sim4
#>            f_dv      td_dv        d_dv lrate_drop_perc
#> 1   0.701028160 0.01351331 -0.47651489            0.00
#> 2   0.663278975 0.07333753 -0.39285858            0.04
#> 3   0.650119216 0.13604230 -0.37299350            0.08
#> 4   0.633004783 0.28672642 -0.30706420            0.12
#> 5   0.556395109 0.31918010 -0.19419062            0.16
#> 6   0.551342130 0.37495150 -0.18118624            0.20
#> 7   0.481881025 0.42253064 -0.11725408            0.24
#> 8   0.416859327 0.48672492  0.02817775            0.28
#> 9   0.442182027 0.54100892  0.02780947            0.32
#> 10  0.352404794 0.55631562  0.07249250            0.36
#> 11  0.334024242 0.60051914  0.13500556            0.40
#> 12  0.295262006 0.62575890  0.19079468            0.44
#> 13  0.307331334 0.70923977  0.21864830            0.48
#> 14  0.271276504 0.66693212  0.20536546            0.52
#> 15  0.191859228 0.70464540  0.26520307            0.56
#> 16  0.188205704 0.73051860  0.28517694            0.60
#> 17  0.172752955 0.74948246  0.34016942            0.64
#> 18  0.100509302 0.76208425  0.38346817            0.68
#> 19  0.129906979 0.77737432  0.34669211            0.72
#> 20  0.151104550 0.76956575  0.33782937            0.76
#> 21  0.073552426 0.81581281  0.42232720            0.80
#> 22  0.030263447 0.79239836  0.46008706            0.84
#> 23  0.023810247 0.81660532  0.44032598            0.88
#> 24  0.005663264 0.78959218  0.43559317            0.92
#> 25  0.064438336 0.83468908  0.44817420            0.96
#> 26 -0.045779705 0.82140950  0.51397210            1.00
```

If the learning rate does not drop much (i.e. *lrate\_drop\_perc*
approaches 1), the influence of total duration (*td\_dv*) on the output
activation is strong, while the influence of frequency (*f\_dv*) is
weak. But if the learning rate drops substantially
(i.e. *lrate\_drop\_perc* approaches 0), the opposite is true.

For more details please check out the vignette in R:

``` r
vignette("passt")
```

Or on CRAN:
<https://cran.r-project.org/web/packages/passt/vignettes/passt.html>

## Issues and Support

If you find any bugs, please use the issue tracker at:

<https://github.com/johannes-titz/passt/issues>

If you need assistance in how to use the package, drop me an e-mail at
johannes at titz.science or johannes.titz at gmail.com

## Contributing

Contributions of any kind are very welcome\! I will sincerely consider
every suggestion on how to improve the code, the documentation and the
presented examples. Even minor things, such as suggestions for better
wording or improving grammar in any part of the package are considered
as valuable contributions.

If you want to make a pull request, please check that you can still
build the package without any errors, warnings or notes. Overall, simply
stick to the R packages book: <https://r-pkgs.org/> and follow the code
style described here: <http://r-pkgs.had.co.nz/r.html#style>

<!-- 
# Citation

Titz, J. (2019). passt: An R implementation of the PASS-T model. R package version 0.1.0. https://CRAN.R-project.org/package=passt

A BibTex entry for LaTeX users is:

@Manual{titz2019, title = {passt: an R implementation of the PASS-T model}, author = {Titz, Johannes}, year = {2019}, note = {R package version 0.1.0}, url = {https://CRAN.R-project.org/package=passt}, }
-->

## References

<div id="refs" class="references">

<div id="ref-Betsch2010">

Betsch, T., Glauer, M., Renkewitz, F., Winkler, I., & Sedlmeier, P.
(2010). Encoding, storage and judgment of experienced frequency and
duration. *Judgment and Decision Making*, *5*, 347–364.

</div>

<div id="ref-Cooper1967">

Cooper, E. H., & Pantle, A. J. (1967). The total-time hypothesis in
verbal learning. *Psychological Bulletin*, *68*, 221–234.
<https://doi.org/10.1037/h0025052>

</div>

<div id="ref-Hintzman1970a">

Hintzman, D. L. (1970). Effects of repetition and exposure duration on
memory. *Journal of Experimental Psychology*, *83*, 435–444.
<https://doi.org/10.1037/h0028865>

</div>

<div id="ref-Hintzman1975">

Hintzman, D. L., Summers, J. J., & Block, R. A. (1975). What causes the
spacing effect? Some effects of repetition, duration, and spacing on
memory for pictures. *Memory & Cognition*, *3*, 287–294.
<https://doi.org/10.3758/BF03212913>

</div>

<div id="ref-Sedlmeier1999">

Sedlmeier, P. (1999). *Improving statistical reasoning: Theoretical
models and practical implications*. Mawah, NJ: Erlbaum.
<https://doi.org/10.4324/9781410601247>

</div>

<div id="ref-Sedlmeier2002a">

Sedlmeier, P. (2002). Associative learning and frequency judgments: The
PASS model. In P. Sedlmeier & T. Betsch (Eds.), *Frequency processing
and cognition* (pp. 137–152). Oxford, England: Oxford University Press.
<https://doi.org/10.1093/acprof:oso/9780198508632.003.0009>

</div>

<div id="ref-Titz2018">

Titz, J., Burkhardt, M., & Sedlmeier, P. (2019). *Asymmetries in
judgments of frequency and duration: A meta-analysis*. Manuscript
submitted for publication.

</div>

<div id="ref-Titz2019">

Titz, J., & Sedlmeier, P. (2019). *Simulating asymmetries in judgments
of frequency and duration: The PASS-T model*. Manuscript submitted for
publication.

</div>

<div id="ref-Winkler2009a">

Winkler, I. (2009). *The processing of frequency and duration* (Doctoral
dissertation, Chemnitz University of Technology). Retrieved from
<https://nbn-resolving.org/urn:nbn:de:bsz:ch1-200900917>

</div>

<div id="ref-Winkler2015">

Winkler, I., Glauer, M., Betsch, T., & Sedlmeier, P. (2015). The impact
of attention on judgments of frequency and duration. *PLoS ONE*, *10*,
e0126974. <https://doi.org/10.1371/journal.pone.0126974>

</div>

</div>


NEWS.md
# passt 0.1.3

* add rmarkdown to suggests field

# passt 0.1.2

* fixed url for Winkler dissertation

# passt 0.1.1

* fix problem in *calc_output_sum* which occurs because the behavior of the 
function *class* changed in the new version of R

# passt 0.1.0

* First Release


tests/testthat.R
library(testthat)
library(passt)

test_check("passt")


tests/testthat/test-simulation.R
library(testthat)
set.seed(20191015)

sim1_weights <- run_sim(patterns = diag(10), frequency = 1:10,
                        duration = 10:1, lrate_onset = 0.05,
                        lrate_drop_time = 2,
                        lrate_drop_perc = 0)$weight_matrix
sim1_weights_before <- as.matrix(read.csv("sim1_weights.csv"))

test_that("sim 1 produces same results as before", {
            expect_equal(sim1_weights, sim1_weights_before,
                         check.attributes = F)
          })

duration <- c(4, 2, 1, 8, 4, 2, 12, 6, 3)
frequency <- c(2, 4, 8, 2, 4, 8, 2, 4, 8)
lrate_drop_perc <- seq(0, 1, 0.04)
sim4 <- lapply(lrate_drop_perc, function(x)
  run_exp(frequency, duration, 0.05, 2, x, diag(9), 10, 0.1))

sim4 <- plyr::ldply(sim4, "data.frame")
sim4 <- cbind(sim4, lrate_drop_perc)

sim4_before <- read.csv("sim4.csv")
test_that("sim 4 produces same results as before", {
            expect_equal(sim4, sim4_before, check.attributes = F)
          })


tests/testthat/test-helper_functions.R
weights_sedlmeier14_2 <- matrix(c(.36, .03, .08, .29,
                                  .27, .53, .29, .15), nrow = 2)

sedlmeier14_2 <- updt_winner_weights(c(1, 0, 0, 1),
                                     weights_sedlmeier14_2,
                                     lrate = .1)

test_that("updating of winner weights works via Sedlmeier1999 fig.
          14.2 (p. 169)", {
            expect_equal(sedlmeier14_2,
                         matrix(c(.374, .072, .243, .311,
                                  .03, .29, .53, .15), byrow = T,
                                nrow = 2))
          })

set.seed(20191016)
test_that("updt_winner_weights work if there are equal output
          activations", {
            expect_equal(updt_winner_weights(c(1, 0),
                                             matrix(rep(0.5, 4),
                                                    nrow = 2),
                                             0.5),
                         matrix(c(0.5, 0.75, 0.50, 0.25), nrow = 2))
          })

test_that("calculating output sum for Sedlmeier fig 14.2 works
          (p. 169)",{
  expect_equal(calc_output_sum(c(1, 0, 0, 1),
                               weights_sedlmeier14_2), .83)
})

test_that("get_attention works", {
  expect_equal(get_attention(10, 1, 2, .5)[[1]], c(1, rep(.5, 9)))
  expect_equal(get_attention(10, 1, 10, .5)[[1]], c(rep(1, 9), .5))
  expect_equal(get_attention(c(10, 5), 1, 6, .2),
               list(c(rep(1, 5), rep(.2, 5)), c(rep(1, 5))))
})


presentation_matrix <- create_pres_matrix(diag(10), 1:10, 10:1,
                                          1, 2, 0.5,
                                          pulses_per_second = 1)[[1]]

test_that("create_pres_matrix works for orthogonal stimuli", {
  expect_equal(colSums(presentation_matrix), c(1:10 * 10:1))
})


example.R
source("pass-t.R")
# very simple example with 9 orthogonal stimuli

# simulation parameter
inputMatrix <- diag(9) # create 9 orthogonal input patterns
presTime <- c(2, 4, 2, 8, 4, 2, 8, 4, 8)
presRate <- c(2, 2, 4, 2, 4, 8, 4, 8, 8)
learningWeight <- 0.1
runs <- 100
AttentionC <- .15 # learning weight falls to AttentionC*learningWeight

# start simulation
MySim <- SimulatePassT(inputMatrix = inputMatrix,
                       learningWeight = learningWeight,
                       presRate = presRate, presTime = presTime,
                       runs=runs, AttentionC=AttentionC, nOutputUnits = 9)

# extract overall activation strength for every input pattern
MySimStrength <- as.data.frame(MySim[[1]])

MySimStrengthSd <- apply(MySimStrength, 2, sd)
MySimStrengthMean <- apply(MySimStrength, 2, mean)

presTimeSum <- presTime*presRate

plot(presTimeSum, MySimStrengthMean, xlab="actual presentation time",
     ylab="Simulation: overall activation",
     main="Activation in simulation as a function of actual presentation time",
     pch=16, bty="l",)
text(presTimeSum, MySimStrengthMean, paste(presRate, "x", presTime, "s",
                                           sep=""), pos=4, cex=0.9)

plot(presRate, MySimStrengthMean, xlab="actual presentation frequency",
     ylab="Simulation: overall activation",
     main="Activation in simulation as a function of actual presentation
     frequency", pch=16, bty="l")
text(presRate, MySimStrengthMean, paste(presRate, "x", presTime, "s", sep=""),
     pos=4, cex=0.9)


empExample.R
source("pass-t.R")

# simulation parameters
presTime <- rep(c(2, 4, 2, 8, 4, 2, 8, 4, 8), 2)
presRate <-rep(c(2, 2, 4, 2, 4, 8, 4, 8, 8), 2)
learningWeight <- 0.01
pulsesPerSecond <- 1
attentionFreq <- c(rep("constant", length(presTime)))
attentionTime <- c(rep("constant", length(presTime)))
inputs <- diag(18)

# Simulation for words
attentionC <- .015

namesSimFreq <- SimulatePassT(inputMatrix = inputs,
                       learningWeight = learningWeight,
                       presRate = presRate, presTime = presTime,
                       runs=500, attentionFunction = attentionFreq,
                       attentionC=attentionC, nOutputUnits = 9)

namesSimFreqStrength <- as.data.frame(namesSimFreq$output)
namesSimFreqStrengthAgg <- aggregate(t(namesSimFreqStrength),
                                     by=list(presRate, presTime*presRate),
                                     mean)[,-c(1,2)]
namesSimFreqJofMean <- apply(namesSimFreqStrengthAgg, 1, mean)

# Simulation for pictures
attentionC <- .1

picsSimFreq <- SimulatePassT(inputMatrix = inputs,
                              learningWeight = learningWeight,
                              presRate = presRate, presTime = presTime,
                              runs=500, attentionFunction = attentionFreq,
                              attentionC=attentionC, nOutputUnits = 9)

picsSimFreqStrength <- as.data.frame(picsSimFreq$output)
picsSimFreqStrengthAgg <- aggregate(t(picsSimFreqStrength),
                                    by=list(presRate, presTime*presRate),
                                    mean)[,-c(1,2)]
picsSimFreqJofMean <- apply(picsSimFreqStrengthAgg, 1, mean)

simJof <- c(namesSimFreqJofMean, picsSimFreqJofMean)

# get empirical data from Isabells experiment 3
if(!exists("isabExp3Judgm.csv")) source("getDataFromIsab.R")
isabExp3 <- read.table("isabExp3Judgm.csv", sep=";", header=T)
# just take names and neutral pics
isabExp3 <- isabExp3[isabExp3$stimulusType %in% c("Namen", "neutrale Bilder"),]

cor(isabExp3$meanJof, simJof)
cor(isabExp3$meanJod, simJof)

# plot ####
mycolors.colorblindsafe <- c("#1b9e77", "#d95f02", "#7570b3")
mycolor <- c(rep(mycolors.colorblindsafe[1], 9), rep(mycolors.colorblindsafe[2], 9))
mypch <- c(rep(16, 9), rep(15, 9))
mycex <- 1.4

pdf("corrEmpSim.pdf", width=10, height=7)
  par(cex=mycex)
  plot(isabExp3$meanJof, simJof,
       xlim=c(min(isabExp3$meanJof), max(isabExp3$meanJof)+0.3),
       xlab="Empirie: Häufigkeitsschätzung",
       ylab="Simulation: Gesamtaktivierung",
       main="Vergleich Simulation und Empirie für Häufigkeitsschätzung",
       col=mycolor, pch=mypch, bty="l")
  # labels for data points
  text(isabExp3$meanJof, simJof, paste(presRate, "x", presTime, "s", sep=""),
       pos=4, cex=0.9)
  # description and correlation
  text(min(isabExp3$meanJof), max(simJof*0.98),
       paste("Zahlen an den Punkten: Häufigkeit x Dauer ",
             "(einzeln)\nKorrelation = ",
             round(cor.test(isabExp3$meanJof, simJof)$estimate,2), sep=""),
       adj=c(0.0, 0.7), offset=0)
  legend("bottomright",
         legend=c("niedrige Aufmerksamkeit", "hohe Aufmerksamkeit"),
         pch=unique(mypch), col=unique(mycolor), bty="n")
dev.off()

R/passt.R
#' Run simulations
#'
#' Runs several simulations and returns output activation for each
#' simulation and each input pattern
#'
#' @param patterns matrix with input patterns, one row is one pattern
#' @param frequency presentation frequency for each pattern in the
#'   matrix
#' @param duration presentation duration for each pattern in the
#'   matrix
#' @param lrate_onset learning rate at the onset of a stimulus
#' @param lrate_drop_time point at which the learning rate drops, must
#'   be lower than duration
#' @param lrate_drop_perc how much the learning rate drops at
#'   lrate_drop_time
#' @param n_runs number of simulations to be run, default is 100
#' @param n_output_units number of output units, defaults to number of
#'   input units
#' @param pulses_per_second how many time steps should be simulated
#'   per second
#' @return list with following elements
#' \itemize{
#'   \item output: the sum of the activation strengths of the
#'     output units for each input pattern
#'   \item weight_matrix: final weight_matrix
#'   \item pres_matrix: presentation matrix
#' }
#' @examples
#' run_sim(diag(10), 1:10, 10:1, 0.05, 2, 0.2)
#' @export
#' @seealso \code{\link{run_exp}}
run_sim <- function(patterns, frequency, duration, lrate_onset,
                    lrate_drop_time, lrate_drop_perc, n_runs = 100,
                    n_output_units = ncol(patterns),
                    pulses_per_second = 1){
  output_activation_sum <- NULL
  for (j in 1:n_runs) {
    n_input_units <- ncol(patterns)
    weight_matrix <- init_weight_mtrx(n_input_units, n_output_units)
    pres <- create_pres_matrix(patterns, frequency, duration,
                               lrate_onset, lrate_drop_time,
                               lrate_drop_perc,
                               pulses_per_second = pulses_per_second)
    # present input i and update weights
    for (i in 1:nrow(pres$input)) {
      weight_matrix <- updt_winner_weights(pres$input[i, ],
                                           weight_matrix,
                                           pres$lrate[i])
    }
    # output for simulation j
    output_activation_sum <- rbind(output_activation_sum,
                                   calc_output_sum(patterns,
                                                   weight_matrix))
  }
  final_list <- list("output" = output_activation_sum,
                    "weight_matrix" = weight_matrix,
                    "pres_matrix" = pres)
  return(final_list)
}

#' Create presentation matrix
#'
#' Taking into account the exposure frequency and exposure duration of
#' the stimuli, this function creates a randomized presentation matrix
#' of all stimuli.
#'
#' @inheritParams run_sim
#' @return list with two lists: the presentation matrix and the
#'   learning weights (both in one random order)
#' @noRd
create_pres_matrix <- function(patterns, frequency, duration,
                               lrate_onset, lrate_drop_time,
                               lrate_drop_perc, pulses_per_second){
  attention <-  get_attention(duration, lrate_onset, lrate_drop_time,
                              lrate_drop_perc, pulses_per_second)
  pres_list <- list()
  final_result <- list()
  # produce list (stimuli) of lists (presentation time) of smallest
  # entity
  for (i in 1:nrow(patterns)) {
    duration_pulse <- duration[i] * pulses_per_second
    pres_list[[i]] <- list(cbind(matrix(rep(patterns[i, ],
                                            duration_pulse),
                                        nrow = duration_pulse,
                                        byrow = TRUE),
                                 attention[[i]]))
    pres_list[[i]] <- lapply(pres_list[i], rep, frequency[i])
  }

  # unlist so that all presentations remain lists, apply unlist two
  # times
  pres_list <- unlist(pres_list, recursive = F)
  pres_list <- unlist(pres_list, recursive = F)

  # randomize presentations
  random_index <- sample(1:length(pres_list), length(pres_list))

  pres_list <- pres_list[random_index]

  # create df
  pres_df <- do.call(rbind, pres_list)

  # split into two lists
  final_result[["input"]] <- pres_df[, 1:(ncol(pres_df) - 1)]
  final_result[["lrate"]] <- pres_df[, ncol(pres_df)]
  return(final_result)
}

#' Initialize weight matrix for competitive learning network
#'
#' Draws weights from a normal distribution and then normalizes these
#' weights, such that the sum of weights going to one output unit is
#' 1.
#'
#' @param n_input_units number of input units
#' @param n_output_units number of output units
#' @param mean mean of normal distribution
#' @param sd standard deviation of normal distribution
#' @return matrix with n_input_units rows and n_output_units columns,
#'   the sum of every column is 1
#' @noRd
init_weight_mtrx <- function(n_input_units, n_output_units,
                             mean = 0.5, sd = 0.005){
  n_weights <- n_output_units * n_input_units
  # initialize random weight mtrx
  weight_mtrx <- matrix(rnorm(n_weights, mean, sd), n_output_units,
                        byrow = TRUE)
  # weights for every output unit (row sum) must equal 1
  weight_mtrx <- prop.table(weight_mtrx, margin = 1)
  return(weight_mtrx)
}

#' Updates weights for competitive learning network algorithm
#'
#' Depending on the input and learning rate, this function takes the
#' weight matrix at step t and returns the weight matrix at step t+1
#'
#' @param input input vector
#' @param weight_matrix weight matrix of network
#' @param lrate learning rate
#' @return new weight matrix for step t + 1
#' @noRd
updt_winner_weights <- function(input, weight_matrix, lrate){
  output <- weight_matrix %*% input
  winner <- which(output == max(output))
  if (length(winner) > 1) {
    warning("more than one winner, random selection")
    winner <- sample(winner, 1)
  }
  delta_w <- lrate * (input / sum(input) - weight_matrix[winner, ])
  weight_matrix[winner, ] <- weight_matrix[winner, ] + delta_w
  return(weight_matrix)
}

#' Calculates sum of activation in output units for input patterns
#'
#' Produces output activations for input patterns, without learning.
#' This is usually used after the learning procedure is completed and
#' one wants to know the reaction of the network to specific input
#' patterns.
#'
#' @param inputs matrix of inputs for which to calculate output
#'   activation
#' @inheritParams updt_winner_weights
#' @return sum of activations of output units for input patterns
#' @noRd
#' @importFrom methods is
calc_output_sum <- function(inputs, weight_matrix){
  if (methods::is(inputs, "matrix")) {
    return(colSums(apply(inputs, 1, function(x) weight_matrix %*% x)))
  }
  if (is.vector(inputs) == TRUE) {
    return(sum(weight_matrix %*% inputs))
  }
}

#' Calculates attention (learning rate) development over time
#'
#' Depending on differnt learning parameters, this function creates
#' the development of the learning rate over time. In the framework of
#' PASS-T, the learning rate can also be seen as the "attention" of
#' the network.
#'
#' @inheritParams run_sim
#'
#' @return list of attention values (learning rate) for every time
#'   pulse for every stimulus
#' @noRd
get_attention <- function(duration, lrate_onset, lrate_drop_time,
                          lrate_drop_perc, pulses_per_second = 1){
  duration_pulses <- duration * pulses_per_second
  lrate_drop_pulse <- lrate_drop_time * pulses_per_second
  low_value <- lrate_onset * lrate_drop_perc
  y <- list()
  for (i in 1:length(duration_pulses)) {
    high <- rep(lrate_onset, lrate_drop_pulse - 1)
    n_low_pulses <- duration_pulses[i] - lrate_drop_pulse + 1
    if (n_low_pulses > 0) {
      low <- rep(low_value, n_low_pulses)
    } else {
      low <- NULL
    }
    y[[i]] <- c(high, low)
  }
  return(y)
}

#' Run simulations and analyze data
#'
#' Runs several simulations and returns correlative effect sizes
#' between the frequency/total duration/single duration of each
#' pattern and the output activation of the network for each pattern,
#' respectively. Comparable to running an empirical experiment in
#' judgments of frequency and duration and analyzing the data.
#'
#' @inheritParams run_sim
#' @param number_of_participants corresponds with number of
#'   simulations run
#' @param cor_noise_sd the amount of noise added to the final
#'   activations of the network, set to 0 if you do not want any noise
#' @export
#' @importFrom dplyr arrange mutate summarize
#' @importFrom tidyr gather
#' @importFrom stats cor rnorm
#' @importFrom rlang .data
#' @examples
#' run_exp(10:1, 1:10, 0.05, 2, 0.2)
#' @seealso  \code{\link{run_sim}}
#' @return data frame with three columns: f_dv, td_dv, t_dv which are
#'   the correlations between the frequency/total duration/single
#'   duration of each pattern and the activation of the network for
#'   each pattern, respectively.
run_exp <- function(frequency, duration, lrate_onset, lrate_drop_time,
                   lrate_drop_perc, patterns = diag(length(duration)),
                   number_of_participants = 100,
                   cor_noise_sd = 0){
  inputs <- diag(length(duration))

  sim_low_a <- run_sim(patterns = patterns, lrate_onset = lrate_onset,
                       lrate_drop_time = 2, frequency = frequency,
                       duration = duration,
                       n_runs = number_of_participants,
                       lrate_drop_perc = lrate_drop_perc,
                       n_output_units = length(duration))

  strength <- as.data.frame(sim_low_a$output)
  # create useful data structure
  d <- cbind(id = 1:nrow(strength), strength)
  # with factor_key = TRUE we can leave the ordering of the columns
  d <- tidyr::gather(d, key = "condition", value = "dv_activation",
                     -.data$id, factor_key = TRUE)
  d <- dplyr::arrange(d, .data$id, .data$condition)
  d <- cbind(d, iv_f = frequency, iv_d = duration,
             iv_total_d = duration * frequency)

  # add noise to dv
  d <- dplyr::mutate(d,
                     dv_activation = .data$dv_activation +
                       stats::rnorm(length(.data$dv_activation),
                                    mean = 0,
                                    sd = cor_noise_sd))
  # calculate effect sizes
  r_contrast <- d %>%
    dplyr::summarize(f_dv = stats::cor(.data$iv_f,
                                       .data$dv_activation),
                     td_dv = stats::cor(.data$iv_total_d,
                                        .data$dv_activation),
                     d_dv = stats::cor(.data$iv_d,
                                       .data$dv_activation))
  r_contrast
}
